VideoMediaSampleRenderer do not display decoded frames that have no duration set.

jyavenard opened the pull request:
#### d33874653d6bf47bba4b9fca76ad6ac8e34b3928
<pre>
VideoMediaSampleRenderer do not display decoded frames that have no duration set.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283677">https://bugs.webkit.org/show_bug.cgi?id=283677</a>
<a href="https://rdar.apple.com/140538938">rdar://140538938</a>

Reviewed by Youenn Fablet.

Do not use CMSampleBuffer&apos;s duration to determine when to schedule the next frame for display or to purge
the decoded queue from late frames.
The decoded sample returned by WebCodec&apos;s VideoDecoder all have a duration set to 0.
Their presentation time is equal to their presentation end time.
When we scheduled the task to purge old frames, it would then attempt to display the earliest frame which would either always be
- too late
- not ready for display yet.

In this change, we now always assume a decoded frame is to be displayed until the next one is available (provided it&apos;s not already late by the time the decoding completed).
We ignore the duration set in the decoded queue&apos;s CMSampleBuffer and instead always use the start time of the next frame
to determine the duration of the previous one.

Covered by existing tests and further tests enabled in a follow-up change (webkit.org/b/282163)

* LayoutTests/media/media-rvfc-paused-mp4-expected.txt: Make seeked event silent as it can happen to be fired after the rVFC promise is resolved.
* LayoutTests/media/media-rvfc-paused-mp4.html:
* LayoutTests/media/media-rvfc-paused-offscreen-mp4-expected.txt:
* LayoutTests/media/media-rvfc-paused-offscreen-mp4.html:
* LayoutTests/media/media-rvfc-paused-offscreen-webm-expected.txt:
* LayoutTests/media/media-rvfc-paused-offscreen-webm.html:
* LayoutTests/media/media-rvfc-paused-webm-expected.txt:
* LayoutTests/media/media-rvfc-paused-webm.html:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h: Add soft-link to CMBufferQueueCallForEachBuffer.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: The timebase and TimerSource should only ever be
accessed atomically together. Change the type to better reflect this by using a std::pair to store both.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::nextDecodedSample const): code was used in multiple place. Move it to a dedicated function.
(WebCore::sampleCallback):
(WebCore::VideoMediaSampleRenderer::nextDecodedSampleEndTime const): Add method that interates the CMBufferQueue
and stops after the 2nd frame.
(WebCore::VideoMediaSampleRenderer::enqueueDecodedSample):
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::clearTimebase):
(WebCore::VideoMediaSampleRenderer::timebaseAndTimerSource const):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::enqueueSample): Notify player that we are ready for more data when clearing the compressed frame queue
(WebCore::VideoMediaSampleRenderer::decodeNextSample): Notify player that we are ready for more data as soon as we pop a frame from the compressed queue.
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::cancelTimer):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::schedulePurgeAndDisplayAtTime):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurgeAndDisplay): Check if we need to reschedule a new purge&amp;display if the latest frame added is more recent than the last one.
Handle the case where we added a new frame with infinite duration (the last frame), in which case we see if it needs to be displayed immediately.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurge): Deleted.

Canonical link: <a href="https://commits.webkit.org/287133@main">https://commits.webkit.org/287133@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/769f8739be46d89cee42c38ff24dabf42f6ff6bd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78478 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31860 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83139 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29743 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80611 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61463 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19378 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51478 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69376 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25240 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28080 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25612 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84505 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5843 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3985 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6004 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17179 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12965 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11354 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5790 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/9181 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5778 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9212 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -111,6 +111,45 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
+}
+
+struct SampleCallbackData {
+    size_t index = 0;
```
How about `bool isFirstSample { true }`?

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -111,6 +111,45 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
+CMTime VideoMediaSampleRenderer::nextDecodedSampleEndTime() const
+{
+    auto sampleCount = decodedSamplesCount();
+    if (!sampleCount)
```
Not strictly needed, maybe CMBufferQueueCallForEachBuffer is already doing it anyway.

jyavenard replied:
yes, it is iterating a CFArray. But it makes it easier to handle the value to return.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -180,12 +219,10 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
     Locker locker { m_lock };
 
-    ASSERT(!m_timebase);
+    ASSERT(!std::get<0>(m_timebaseAndTimerSource));
```
This is not really readable, how about `ASSERT(!m_timebaseAndTimerSource.first)`?

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -194,33 +231,42 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
-    m_timerSource = nullptr;
+    PAL::CMTimebaseRemoveTimerDispatchSource(timebase.get(), timerSource.get());
+    dispatch_source_cancel(timerSource.get());
+    m_timebaseAndTimerSource = { nullptr, nullptr };
```
Would `{ }` work here?

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
```diff
@@ -97,6 +97,8 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
 
     void clearTimebase();
+    using TimebaseAndTimerSource = std::pair<RetainPtr<CMTimebaseRef>, OSObjectPtr<dispatch_source_t>>;
+    TimebaseAndTimerSource timebaseAndTimerSource() const;
```
I know this is preexisting but it is bit odd to have Timebase and TimerSource...

jyavenard replied:
they are often accessed together really, and theoretically, they should be modified together atomically. hence I think combining them in the data type makes sense

jyavenard replied:
I understand your comment now. The CoreMedia object is named `CMTimebase` so it's consistent with the inconsistency of CoreMedia methods naming :)

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -471,8 +510,10 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
-    Locker locker { m_lock };
-    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+    auto [timebase, timerSource] = timebaseAndTimerSource();
+    if (!timebase || !timerSource)
```
In other part of the code, we just do a `if(!timebase)` which is probably good enough.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -482,32 +523,23 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 
-    while (RetainPtr nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()))) {
-        CMTime duration = PAL::CMSampleBufferGetOutputDuration(nextSample.get());
+    while (RetainPtr nextSample = this->nextDecodedSample()) {
```
Why `this->`?
Here and elsewhere in this patch like below.

jyavenard replied:
that method used to be called `nextSample()` I didn't update the calling point after I renamed the method to be more consistent with the rest.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -536,18 +567,11 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
-    if (PAL::CMTimeCompare(*m_nextScheduledPurge, presentationEndTime) > 0)
-        purgeDecodedSampleQueue(ShouldReschedule::Yes);
+    auto [timebase, timerSource] = timebaseAndTimerSource();
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(timebase.get(), timerSource.get(), presentationEndTime, 0);
```
Should we set `m_nextScheduledPurge` to `presentationEndTime`?
And check it in the call of `CMTimebaseSetTimerDispatchSourceNextFireTime` in `purgeDecodedSampleQueue`?

Or maybe call an updated version of `maybeReschedulePurge` in `purgeDecodedSampleQueue`.
That way, only one method would be responsible to call `CMTimebaseSetTimerDispatchSourceNextFireTime`.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -390,59 +422,46 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
-            LOG(Media, "maybeQueueFrameForDisplay: currentTime:%f presentationTime:%f endTime:%f TooEarly", PAL::CMTimeGetSeconds(currentTime), PAL::CMTimeGetSeconds(presentationTime), PAL::CMTimeGetSeconds(presentationEndTime));
-            return DecodedFrameResult::TooEarly;
+        if (!m_forceLateSampleToBeDisplayed && m_isDisplayingSample && m_lastDisplayedTime) {
+            auto comparisonResult = PAL::CMTimeCompare(presentationTime, *m_lastDisplayedTime);
```
`comparisonResult ` is not really needed.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -180,47 +217,43 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
     Locker locker { m_lock };
 
-    if (!m_timebase)
+    auto [timebase, timerSource] = m_timebaseAndTimerSource;
```
How about using `std::exchange(m_timebaseAndTimerSource, { })` here?

jyavenard replied:
ok

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -390,59 +422,46 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 }
 
-VideoMediaSampleRenderer::DecodedFrameResult VideoMediaSampleRenderer::maybeQueueFrameForDisplay(CMSampleBufferRef sample, FlushId flushId)
+VideoMediaSampleRenderer::DecodedFrameResult VideoMediaSampleRenderer::maybeQueueFrameForDisplay(const CMTime& currentTime, CMSampleBufferRef sample, FlushId flushId)
```
I would be tempted to pass CMTime by value.

jyavenard replied:
The struct is 20 bytes long.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -390,59 +422,46 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
-    if (RetainPtr timebase = this->timebase()) {
-        if (m_lastDisplayedTime) {
-            auto comparisonResult = PAL::CMTimeCompare(presentationTime, *m_lastDisplayedTime);
+    if (CMTIME_IS_VALID(currentTime)) {
```
Is it possible to have an invalid currentTime?

jyavenard replied:
hmm, you're right, this is no longer possible in this revision. Previously there was handling for using a VideoMediaSampleRenderer without a CMtimebase set and it would immediately display the frame once decoded. I've lost this option along the way:(

But considering it's never used that way, it may not be a bad thing anyway.

jyavenard committed:
VideoMediaSampleRenderer do not display decoded frames that have no duration set.
https://bugs.webkit.org/show_bug.cgi?id=283677
rdar://140538938

Reviewed by Youenn Fablet.

Do not use CMSampleBuffer's duration to determine when to schedule the next frame for display or to purge
the decoded queue from late frames.
The decoded sample returned by WebCodec's VideoDecoder all have a duration set to 0.
Their presentation time is equal to their presentation end time.
When we scheduled the task to purge old frames, it would then attempt to display the earliest frame which would either always be
- too late
- not ready for display yet.

In this change, we now always assume a decoded frame is to be displayed until the next one is available (provided it's not already late by the time the decoding completed).
We ignore the duration set in the decoded queue's CMSampleBuffer and instead always use the start time of the next frame
to determine the duration of the previous one.

Covered by existing tests and further tests enabled in a follow-up change (webkit.org/b/282163)

* LayoutTests/media/media-rvfc-paused-mp4-expected.txt: Make seeked event silent as it can happen to be fired after the rVFC promise is resolved.
* LayoutTests/media/media-rvfc-paused-mp4.html:
* LayoutTests/media/media-rvfc-paused-offscreen-mp4-expected.txt:
* LayoutTests/media/media-rvfc-paused-offscreen-mp4.html:
* LayoutTests/media/media-rvfc-paused-offscreen-webm-expected.txt:
* LayoutTests/media/media-rvfc-paused-offscreen-webm.html:
* LayoutTests/media/media-rvfc-paused-webm-expected.txt:
* LayoutTests/media/media-rvfc-paused-webm.html:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h: Add soft-link to CMBufferQueueCallForEachBuffer.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: The timebase and TimerSource should only ever be
accessed atomically together. Change the type to better reflect this by using a std::pair to store both.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::nextDecodedSample const): code was used in multiple place. Move it to a dedicated function.
(WebCore::sampleCallback):
(WebCore::VideoMediaSampleRenderer::nextDecodedSampleEndTime const): Add method that interates the CMBufferQueue
and stops after the 2nd frame.
(WebCore::VideoMediaSampleRenderer::enqueueDecodedSample):
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::clearTimebase):
(WebCore::VideoMediaSampleRenderer::timebaseAndTimerSource const):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::enqueueSample): Notify player that we are ready for more data when clearing the compressed frame queue
(WebCore::VideoMediaSampleRenderer::decodeNextSample): Notify player that we are ready for more data as soon as we pop a frame from the compressed queue.
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::cancelTimer):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::schedulePurgeAndDisplayAtTime):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurgeAndDisplay): Check if we need to reschedule a new purge&display if the latest frame added is more recent than the last one.
Handle the case where we added a new frame with infinite duration (the last frame), in which case we see if it needs to be displayed immediately.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurge): Deleted.

Canonical link: https://commits.webkit.org/287133@main
